closed/src/java.base/share/native/libjncrypto/NativeCrypto.c

```diff
@@ -474,12 +566,262 @@ JNIEXPORT jlong JNICALL Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+                previousVersion = tempVersion;
+                prevResult = result;
+            }
+        }
+    }
+
+    /* If we reach here, it means that none of the non-generic libraries
+     * where found. However, a generic one might have been found in the
+     * process and, if so, it will be in the prevResult variable.
+     */
+    return prevResult;
+}
+
+/*
+ * Class:     jdk_crypto_jniprovider_NativeCrypto
+ * Method:    loadCrypto
+ * Signature: (ZLjava/lang/String;Ljava/lang/String;)J
+ */
+JNIEXPORT jlong JNICALL
+Java_jdk_crypto_jniprovider_NativeCrypto_loadCrypto
+  (JNIEnv * env, jobject jobj, jboolean traceEnabled, jstring jlibname, jstring jhomepath)
+{
+    const char *clibname = "";
+    const char *chomepath = "";
+    jlong ossl_ver = 0;
+
+    if (NULL != jlibname) {
+        clibname = (*env)->GetStringUTFChars(env, jlibname, NULL) ;
+        if (clibname[0] == '\0') {
+            if (traceEnabled) {
+                fprintf(stderr, "The jdk.native.openssl.lib property is not set.\n");
+                fflush(stderr);
+            }
+        } else {
+            crypto_library = load_crypto_library(traceEnabled, clibname);
+            if (NULL == crypto_library) {
+                if (traceEnabled) {
+                    fprintf(stderr, "OpenSSL library specified in jdk.openssl.lib couldn't be loaded.\n");
+                    fflush(stderr);
+                }
+                return -1;
+            }
+        }
+    }
+
+    if (NULL != jhomepath) {
+        chomepath = (*env)->GetStringUTFChars(env, jhomepath, NULL);
+    }
+
+    /* If the jdk.native.openssl.lib property was not set, attempt
+     * to find an OpenSSL library from java.home or OS Library path.
+     */
+    if (NULL == crypto_library){
+        crypto_library = find_crypto_library(traceEnabled, chomepath);
+    }
+
+    (*env)->ReleaseStringUTFChars(env, jlibname, clibname);
+    (*env)->ReleaseStringUTFChars(env, jhomepath, chomepath);
+
+    /* If an OpenSSL library was not loaded from any of the potential
+     * sources, fail loading native crypto.
+     */
+    if (NULL == crypto_library) {
+        if (traceEnabled) {
+            fprintf(stderr, "FAILED TO LOAD OPENSSL CRYPTO LIBRARY\n");
+            fflush(stderr);
+        }
+        return -1;
+    }
+
+    log_crypto_library_path(traceEnabled, crypto_library, "OpenSSL to be used was loaded from");
+    ossl_ver = get_crypto_library_version(traceEnabled, crypto_library, "Version of OpenSSL library that is used");
+
     /* Load the function symbols for OpenSSL errors. */
     OSSL_error_string_n = (OSSL_error_string_n_t*)find_crypto_symbol(crypto_library, "ERR_error_string_n");
     OSSL_error_string = (OSSL_error_string_t*)find_crypto_symbol(crypto_library, "ERR_error_string");
     OSSL_get_error = (OSSL_get_error_t*)find_crypto_symbol(crypto_library, "ERR_get_error");
 
-    /* Load Threading routines for OpenSSL 1.0.2 */
+    /* Load Threading routines for OpenSSL 1.0.2. */
```

Comment:
There are a number of C++-style comments below: could you please fix them while you're here? See lines 2356, 2872, 2904, etc.

Reply:
Updated all those comments to C-style.